src/mozilla_sec_eia/models/sec10k/ex_21/ex21_validation_helpers.py

```diff
@@ -1,13 +1,132 @@
+                value_col=col,
+            )
+
+    jaccard_df = pd.DataFrame.from_dict(jaccard_dict, orient="index").reset_index()
+    prec_recall_df = pd.DataFrame.from_dict(
+        table_metrics_dict, orient="index"
+    ).reset_index()
+
+    return (
+        jaccard_df,
+        prec_recall_df,
+        pd.DataFrame({"filename": incorrect_files}),
+        {
+            "table_accuracy": n_equal / n_files,
+            "avg_subsidiary_jaccard_sim": jaccard_df["subsidiary"].sum() / n_files,
+            "avg_location_jaccard_sim": jaccard_df["loc"].sum() / n_files,
+            "avg_own_per_jaccard_sim": jaccard_df["own_per"].sum() / n_files,
+            "avg_subsidiary_precision": prec_recall_df["subsidiary_precision"].sum()
+            / n_files,
+            "avg_location_precision": prec_recall_df["loc_precision"].sum() / n_files,
+            "avg_own_per_precision": prec_recall_df["own_per_precision"].sum()
+            / n_files,
+            "avg_subsidiary_recall": prec_recall_df["subsidiary_recall"].sum()
+            / n_files,
+            "avg_location_recall": prec_recall_df["loc_recall"].sum() / n_files,
+            "avg_own_per_recall": prec_recall_df["own_per_recall"].sum() / n_files,
+        },
+    )
+
+
+def clean_extracted_df(extracted_df):
+    """Perform basic cleaning on a dataframe extracted from an Ex. 21."""
+    if extracted_df.empty:
+        return extracted_df
+    if "row" in extracted_df.columns:
+        extracted_df = extracted_df.drop(columns=["row"])
+    extracted_df["subsidiary"] = extracted_df["subsidiary"].str.strip().str.lower()
+    # strip special chars from the start and end of the string
+    extracted_df["subsidiary"] = extracted_df["subsidiary"].str.replace(
+        r"^[^\w&\s]+|[^\w&\s]+$", "", regex=True
+    )
+    if "loc" in extracted_df.columns:
+        extracted_df["loc"] = extracted_df["loc"].str.strip().str.lower()
+        extracted_df["loc"] = extracted_df["loc"].str.replace(
+            r"[^a-zA-Z&,\s]", "", regex=True
+        )
+    if "own_per" in extracted_df.columns:
+        # remove special chars and letters
+        extracted_df["own_per"] = extracted_df["own_per"].str.replace(
+            r"[^\d.]", "", regex=True
+        )
+        # Find values with multiple decimal points
+        extracted_df["own_per"] = extracted_df["own_per"].str.replace(
+            r"(\d*\.\d+)\..*", r"\1", regex=True
+        )
+        extracted_df["own_per"] = extracted_df["own_per"].replace("", np.nan)
+        extracted_df["own_per"] = extracted_df["own_per"].astype(
+            "float64", errors="ignore"
+        )
+    # drop rows that have a null subsidiary value
+    extracted_df = extracted_df.dropna(subset="subsidiary")
+    return extracted_df
+
+
 def clean_ex21_validation_set(validation_df: pd.DataFrame):
     """Clean Ex. 21 validation data to match extracted format."""
+    validation_df = remove_paragraph_tables_from_validation_data(validation_df)
```

Comment:
I added this, to automatically remove paragraph layout docs from the validation data (even though the current version of the validation data already has them removed). I figured it couldn't hurt, but you can take it out if you want.